File: shell_wrapper/shell_serialization.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "shell_wrapper/shell_serialization.h"

#include <cstdint>
#include <memory>
#include <string>
#include <utility>

#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "include/cxx.h"
#include "shell_encryption/rns/rns_serialization.pb.h"
#include "shell_wrapper/shell_types.h"
#include "shell_wrapper/shell_types.rs.h"
#include "shell_wrapper/status.h"
#include "shell_wrapper/status.rs.h"

using secure_aggregation::MakeFfiStatus;

FfiStatus SerializeRnsPolynomialToBytes(const RnsPolynomialWrapper* poly,
                                        ModuliWrapper moduli,
                                        std::unique_ptr<std::string>& out) {
  if (poly == nullptr || poly->ptr == nullptr || moduli.moduli == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        "All pointer arguments and their wrapped pointers must be non-null."));
  }
  auto serialized = poly->ptr->Serialize({moduli.moduli, moduli.len});
  if (!serialized.ok()) {
    return MakeFfiStatus(serialized.status());
  }
  std::string buffer;
  if (!serialized->SerializeToString(&buffer)) {
    return MakeFfiStatus(
        absl::InternalError("Failed to serialize RNS polynomial to string."));
  }
  out = std::make_unique<std::string>(std::move(buffer));
  return MakeFfiStatus();
}

FfiStatus DeserializeRnsPolynomialFromBytes(
    rust::Slice<const uint8_t> serialized_poly, ModuliWrapper moduli,
    RnsPolynomialWrapper* out) {
  if (out == nullptr || out->ptr == nullptr || moduli.moduli == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        "All pointer arguments and their wrapped pointers must be non-null."));
  }
  rlwe::SerializedRnsPolynomial serialized_poly_proto;
  if (!serialized_poly_proto.ParseFromString(
          ToAbslStringView(serialized_poly))) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        "Failed to parse serialized RNS polynomial"));
  }
  auto poly = secure_aggregation::RnsPolynomial::Deserialize(
      serialized_poly_proto, {moduli.moduli, moduli.len});
  if (!poly.ok()) {
    return MakeFfiStatus(poly.status());
  }
  out->ptr = std::make_unique<secure_aggregation::RnsPolynomial>(
      std::move(poly.value()));
  return MakeFfiStatus();
}

File: shell_wrapper/shell_serialization.rs
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Rust wrapper for serialization support of SHELL types.

use protobuf::prelude::*;
use rns_serialization_rust_proto::SerializedRnsPolynomial;
use shell_types::{create_empty_rns_polynomial, Moduli, RnsPolynomial};
use status::{StatusError, StatusErrorCode};

#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("shell_wrapper/shell_serialization.h");
        include!("shell_wrapper/shell_types.h");

        type FfiStatus = shell_types::ffi::FfiStatus;
        type ModuliWrapper = shell_types::ffi::ModuliWrapper;
        type RnsPolynomialWrapper = shell_types::ffi::RnsPolynomialWrapper;

        pub unsafe fn SerializeRnsPolynomialToBytes(
            poly: *const RnsPolynomialWrapper,
            moduli: ModuliWrapper,
            out: &mut UniquePtr<CxxString>,
        ) -> FfiStatus;

        pub unsafe fn DeserializeRnsPolynomialFromBytes(
            serialized_poly: &[u8],
            moduli: ModuliWrapper,
            out: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;

    }
}

use status::rust_status_from_cpp;

// Serialize a RnsPolynomial to a SerializedRnsPolynomial proto.
pub fn rns_polynomial_to_proto(
    poly: &RnsPolynomial,
    moduli: &Moduli,
) -> Result<SerializedRnsPolynomial, status::StatusError> {
    let mut out = cxx::UniquePtr::null();
    // SAFETY: No lifetime constraints (no references are kept by the C++ function).
    // `SerializeRnsPolynomialToBytes` allocates a C++ string to write the proto bytes to, and assigns
    // the string to `out`.
    rust_status_from_cpp(unsafe {
        ffi::SerializeRnsPolynomialToBytes(poly, moduli.moduli, &mut out)
    })?;
    SerializedRnsPolynomial::parse(out.as_bytes()).map_err(|parse_error| {
        StatusError::new_with_current_location(
            StatusErrorCode::Internal,
            format!("{parse_error:?}"),
        )
    })
}

// Deserialize a SerializedRnsPolynomial proto to a RnsPolynomial.
pub fn rns_polynomial_from_proto(
    serialized: SerializedRnsPolynomial,
    moduli: &Moduli,
) -> Result<RnsPolynomial, status::StatusError> {
    let serialized_bytes = serialized.serialize().map_err(|serialize_error| {
        StatusError::new_with_current_location(
            StatusErrorCode::Internal,
            format!("{serialize_error:?}"),
        )
    })?;

    // SAFETY: No lifetime constraints (`create_empty_rns_polynomial` creates and returns an empty
    // C++ object).
    let mut poly = unsafe { create_empty_rns_polynomial() };

    // SAFETY: No lifetime constraints (no references are kept by the C++ function).
    // `DeserializeRnsPolynomialFromBytes` allocates a C++ RnsPolynomial object and assigns it to `poly`.
    rust_status_from_cpp(unsafe {
        ffi::DeserializeRnsPolynomialFromBytes(&serialized_bytes, moduli.moduli, &mut poly)
    })?;
    Ok(poly)
}
